pilot/pkg/model/push_context.go

```diff
@@ -763,6 +779,50 @@ func (ps *PushContext) initServiceAccounts(env *Environment, services []*Service
 	}
 }
 
+// Caches list of authentication policies
+func (ps *PushContext) initAuthNPolicies(env *Environment) error {
+	authNPolicies, err := env.List(schemas.AuthenticationPolicy.Type, NamespaceAll)
```

Comment:
I actually like the previous implementations. It stays with the existing behavior of hostname -> policy look up. If we want to do it this way, probably need to change the other function (`AuthenticationPolicyForWorkload` under `model/config.go`) as well.

Reply:
@diemtvu the issue with the previous implementation is that it requires a configstore lookup call for every service/port combination - which would defeat the purpose of the cache right?  Meaning we would just be moving all the lookup calls currently in the XDS generation logic to the init of push_context?

Reply:
I see, I thought config store already have some layer of caching. For this, can you refactor into few functions like these:

```
# return mesh policy and all namespaces policies
GetAllAuthNPolicies(..) (*Config, []Config) {...}

# return the policy for servic-port, given mesh policy and all namespace-policies.
GetAuthNPolicyForService(service *Service, l labels.Instance, port *Port, *Config, [] Config) {...}
```

Then, the init function can repeatedly call GetAuthNPolicyForService for each service-port as you did before. 

The (model) `AuthenticationPolicyForWorkload` can also reuse these two so we don't have any discrepancy.

Reply:
@diemtvu I had a similar train of thought already in progress.  Let me know if you approve of the current implementation